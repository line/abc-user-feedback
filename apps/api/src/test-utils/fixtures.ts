/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { faker } from '@faker-js/faker';
import * as bcrypt from 'bcrypt';

import {
  FieldFormatEnum,
  FieldStatusEnum,
  FieldTypeEnum,
  isSelectFieldFormat,
} from '@/common/enums';
import type { ChannelEntity } from '@/domains/admin/channel/channel/channel.entity';
import type {
  CreateFieldDto,
  ReplaceFieldDto,
} from '@/domains/admin/channel/field/dtos';
import type { FieldEntity } from '@/domains/admin/channel/field/field.entity';
import type { FeedbackEntity } from '@/domains/admin/feedback/feedback.entity';
import type { CreateIssueDto } from '@/domains/admin/project/issue/dtos';
import type { TenantEntity } from '@/domains/admin/tenant/tenant.entity';
import {
  SignUpMethodEnum,
  UserStateEnum,
  UserTypeEnum,
} from '@/domains/admin/user/entities/enums';
import type { UserEntity } from '@/domains/admin/user/entities/user.entity';

export const createFieldEntity = (input: Partial<CreateFieldDto>) => {
  const format = input?.format ?? getRandomEnumValue(FieldFormatEnum);
  const type = input?.type ?? getRandomEnumValue(FieldTypeEnum);
  const status = input?.status ?? getRandomEnumValue(FieldStatusEnum);
  return {
    name: faker.string.alphanumeric(20),
    description: faker.lorem.lines(2),
    format,
    type,
    status,
    options:
      format === FieldFormatEnum.select
        ? getRandomOptionEntities().sort(optionSort)
        : undefined,
    ...input,
  };
};
export const createFieldDto = (input: Partial<CreateFieldDto>) => {
  const format = input?.format ?? getRandomEnumValue(FieldFormatEnum);
  const type = input?.type ?? getRandomEnumValue(FieldTypeEnum);
  const status = input?.status ?? getRandomEnumValue(FieldStatusEnum);
  return {
    name: faker.string.alphanumeric(20),
    key: faker.string.alphanumeric(20),
    description: faker.lorem.lines(2),
    format,
    type,
    status,
    options: isSelectFieldFormat(format)
      ? getRandomOptionDtos().sort(optionSort)
      : undefined,
    ...input,
  };
};
export const updateFieldDto = (input: Partial<ReplaceFieldDto>) => {
  return {
    id: faker.number.int(),
    ...createFieldDto(input),
  };
};

export const createIssueDto = (input: Partial<CreateIssueDto>) => {
  return {
    name: faker.string.alphanumeric(20),
    ...input,
  };
};

export const getRandomValue = (
  format: FieldFormatEnum,
  options?: { id: number; name: string; key: string }[],
) => {
  switch (format) {
    case FieldFormatEnum.text:
      return faker.string.sample();
    case FieldFormatEnum.keyword:
      return faker.string.sample();
    case FieldFormatEnum.number:
      return faker.number.int();
    case FieldFormatEnum.boolean:
      return faker.datatype.boolean();
    case FieldFormatEnum.select:
      return options.length === 0
        ? undefined
        : options[faker.number.int({ min: 0, max: options.length - 1 })].key;
    case FieldFormatEnum.multiSelect:
      return options.length === 0
        ? []
        : faker.helpers
            .shuffle(options)
            .slice(0, faker.number.int({ min: 0, max: options.length - 1 }))
            .map((option) => option.key);
    case FieldFormatEnum.date:
      return faker.date.anytime().toISOString();
    case FieldFormatEnum.image:
      return faker.string.sample();
    default:
      throw new Error('Invalid field type ');
  }
};

const getRandomOptionEntities = () => {
  const length = faker.number.int(10);
  return Array.from({ length }).map(() => ({
    id: faker.number.int(),
    name: faker.string.sample(),
  }));
};
const getRandomOptionDtos = () => {
  const length = faker.number.int(10);
  return Array.from({ length }).map(() => {
    const randomValue = faker.string.sample();
    return {
      id: faker.number.int(),
      name: randomValue,
      key: randomValue,
    };
  });
};

export const getRandomEnumValue = <T>(anEnum: T): T[keyof T] => {
  const enumValues = Object.keys(anEnum) as Array<keyof T>;
  const randomIndex = faker.number.int(enumValues.length - 1);
  const randomEnumKey = enumValues[randomIndex];
  return anEnum[randomEnumKey];
};
export const getRandomEnumValues = <T>(anEnum: T): T[keyof T][] => {
  const enumValues = Object.values(anEnum) as Array<keyof T>;
  return faker.helpers.arrayElements(enumValues) as T[keyof T][];
};

export const optionSort = (a, b) =>
  a.name < b.name ? -1 : a.name > b.name ? 1 : 0;

export const fieldsFixture = Object.values(FieldFormatEnum).flatMap((format) =>
  Object.values(FieldTypeEnum).flatMap((type) =>
    Object.values(FieldStatusEnum).flatMap((status) => ({
      id: faker.number.int(),
      ...createFieldDto({
        format,
        type,
        status,
      }),
    })),
  ),
) as FieldEntity[];

export const feedbackDataFixture = fieldsFixture.reduce((prev, curr) => {
  if (curr.type === FieldTypeEnum.ADMIN) return prev;
  if (curr.status === FieldStatusEnum.INACTIVE) return prev;
  const value = getRandomValue(curr.format, curr.options);
  return {
    ...prev,
    [curr.key]: value,
  };
}, {});

export const feedbackFixture = {
  id: faker.number.int(),
  rawData: feedbackDataFixture,
  additionalData: {},
  createdAt: faker.date.past(),
  updatedAt: faker.date.past(),
  channel: {
    id: faker.number.int(),
    name: faker.string.sample(),
    description: faker.lorem.lines(2),
    imageConfig: null,
    createdAt: faker.date.past(),
    updatedAt: faker.date.past(),
  } as ChannelEntity,
  issues: [],
} as FeedbackEntity;

export const passwordFixture = faker.internet.password();

export const emailFixture = faker.internet.email();

export const userFixture = {
  id: faker.number.int(),
  email: emailFixture,
  name: faker.string.sample(),
  department: faker.string.sample(),
  state: getRandomEnumValue(UserStateEnum),
  hashPassword: bcrypt.hashSync(passwordFixture, 0),
  type: getRandomEnumValue(UserTypeEnum),
  signUpMethod: getRandomEnumValue(SignUpMethodEnum),
} as UserEntity;

export const tenantFixture = {
  id: faker.number.int(),
  siteName: faker.string.sample(),
  description: faker.lorem.lines(2),
  useEmail: faker.datatype.boolean(),
  isPrivate: faker.datatype.boolean(),
  isRestrictDomain: faker.datatype.boolean(),
  allowDomains: [],
  useOAuth: faker.datatype.boolean(),
  oauthConfig: null,
  createdAt: faker.date.past(),
  updatedAt: faker.date.past(),
} as TenantEntity;
