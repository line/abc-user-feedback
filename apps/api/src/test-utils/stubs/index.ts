/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
export { ApiKeyRepositoryStub } from './api-key-repository.stub';
export { ChannelRepositoryStub } from './channel-repository.stub';
export { CodeRepositoryStub } from './code-repository.stub';
export { EventRepositoryStub } from './event-repository.stub';
export { FeedbackRepositoryStub } from './feedback-repository.stub';
export { FieldRepositoryStub } from './field-repository.stub';
export { IssueRepositoryStub } from './issue-repository.stub';
export { ProjectRepositoryStub } from './project-repository.stub';
export { RoleRepositoryStub } from './role-repository.stub';
export { TenantRepositoryStub } from './tenant-repository.stub';
export { UserRepositoryStub } from './user-repository.stub';
export { WebhookRepositoryStub } from './webhook-repository.stub';
