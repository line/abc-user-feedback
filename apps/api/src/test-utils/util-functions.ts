/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { faker } from '@faker-js/faker';
import type { InjectionToken, Provider } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import type { DataSource, Repository } from 'typeorm';
import { initializeTransactionalContext } from 'typeorm-transactional';

import { jwtConfig, jwtConfigSchema } from '@/configs/jwt.config';
import {
  opensearchConfig,
  opensearchConfigSchema,
} from '@/configs/opensearch.config';
import { smtpConfig, smtpConfigSchema } from '@/configs/smtp.config';
import type { AuthService } from '@/domains/admin/auth/auth.service';
import { UserDto } from '@/domains/admin/user/dtos';
import { UserStateEnum } from '@/domains/admin/user/entities/enums';
import { UserEntity } from '@/domains/admin/user/entities/user.entity';

initializeTransactionalContext();

export const getMockProvider = (
  injectToken: InjectionToken,
  factory: unknown,
): Provider => ({ provide: injectToken, useFactory: () => factory });

export const TestConfig = ConfigModule.forRoot({
  load: [smtpConfig, jwtConfig, opensearchConfig],
  envFilePath: '.env.test',
  validationSchema: smtpConfigSchema
    .concat(jwtConfigSchema)
    .concat(opensearchConfigSchema),
});

export const MockDataSource = {
  initialize: jest.fn(),
};

export const getRandomEnumValue = <T>(anEnum: T): T[keyof T] => {
  const enumValues = Object.keys(anEnum) as Array<keyof T>;
  const randomIndex = faker.number.int(enumValues.length - 1);
  const randomEnumKey = enumValues[randomIndex];
  return anEnum[randomEnumKey];
};
export const getRandomEnumValues = <T>(anEnum: T): T[keyof T][] => {
  const enumValues = Object.values(anEnum) as Array<keyof T>;
  return faker.helpers.arrayElements(enumValues) as T[keyof T][];
};

export const clearEntities = async (repos: Repository<any>[]) => {
  for (const repo of repos) {
    await repo.query('set foreign_key_checks = 0');
    await repo.clear();
    await repo.query('set foreign_key_checks = 1');
  }
};

export const signInTestUser = async (
  dataSource: DataSource,
  authService: AuthService,
) => {
  const userRepo = dataSource.getRepository(UserEntity);
  const user = await userRepo.save({
    email: faker.internet.email(),
    state: UserStateEnum.Active,
    hashPassword: faker.internet.password(),
  });
  return { jwt: await authService.signIn(UserDto.transform(user)), user };
};

export const DEFAULT_FIELD_COUNT = 2;

export const createQueryBuilder: any = {
  setFindOptions: () => jest.fn().mockImplementation(() => createQueryBuilder),
  select: () => createQueryBuilder,
  innerJoin: () => createQueryBuilder,
  leftJoin: () => createQueryBuilder,
  where: () => createQueryBuilder,
  andWhere: () => createQueryBuilder,
  groupBy: () => createQueryBuilder,
  addOrderBy: () => createQueryBuilder,
  getRawMany: () => createQueryBuilder,
  insert: () => createQueryBuilder,
  values: () => createQueryBuilder,
  orUpdate: () => createQueryBuilder,
  updateEntity: () => createQueryBuilder,
  execute: () => createQueryBuilder,
  offset: () => createQueryBuilder,
  limit: () => createQueryBuilder,
  getMany: () => createQueryBuilder,
  getCount: () => createQueryBuilder,
};

export const mockRepository = () => ({
  findOneBy: jest.fn(),
  findOne: jest.fn(),
  find: jest.fn(),
  findBy: jest.fn(),
  findAndCount: jest.fn(),
  findAndCountBy: jest.fn(),
  save: jest.fn(),
  create: jest.fn(),
  update: jest.fn(),
  upsert: jest.fn(),
  count: jest.fn(),
  remove: jest.fn(),
  createQueryBuilder: jest.fn(() => createQueryBuilder),
  query: jest.fn(),
});

export const MockOpensearchRepository = {
  createIndex: jest.fn(),
  deleteIndex: jest.fn(),
  putMappings: jest.fn(),
  createData: jest.fn(),
  getData: jest.fn(),
  updateData: jest.fn(),
  getTotal: jest.fn(),
};
