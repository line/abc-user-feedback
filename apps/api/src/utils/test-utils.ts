/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { faker } from '@faker-js/faker';
import { MailerModule } from '@nestjs-modules/mailer';
import { InjectionToken, Provider } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { TypeOrmModule } from '@nestjs/typeorm';
import { join } from 'path';
import { DataSource, Repository } from 'typeorm';
import { SnakeNamingStrategy } from 'typeorm-naming-strategies';
import {
  addTransactionalDataSource,
  initializeTransactionalContext,
} from 'typeorm-transactional';

import { jwtConfig, jwtConfigSchema } from '@/configs/jwt.config';
import { mySqlConfigSchema, mysqlConfig } from '@/configs/mysql.config';
import { smtpConfig, smtpConfigSchema } from '@/configs/smtp.config';
import { AuthService } from '@/domains/auth/auth.service';
import { UserDto } from '@/domains/user/dtos';
import { UserStateEnum } from '@/domains/user/entities/enums';
import { UserEntity } from '@/domains/user/entities/user.entity';
import { ConfigServiceType } from '@/types/config-service.type';

initializeTransactionalContext();

export const getMockProvider = (
  injectToken: InjectionToken,
  factory: unknown,
): Provider => ({ provide: injectToken, useFactory: () => factory });

export const TestConfig = ConfigModule.forRoot({
  load: [smtpConfig, jwtConfig, mysqlConfig],
  envFilePath: '.env.test',
  validate: (config) => ({
    ...smtpConfigSchema.validateSync(config),
    ...jwtConfigSchema.validateSync(config),
    ...mySqlConfigSchema.validateSync(config),
  }),
});

export const TestMailConfigModule = MailerModule.forRoot({
  transport: { url: '' },
});

export const TestTypeOrmConfig = TypeOrmModule.forRootAsync({
  imports: [ConfigModule],
  inject: [ConfigService],
  dataSourceFactory: async (options) => {
    const datasource = await new DataSource(options).initialize();
    return addTransactionalDataSource(datasource);
  },
  useFactory: (configService: ConfigService<ConfigServiceType>) => {
    const { main_url, sub_urls } = configService.get('mysql', {
      infer: true,
    });
    return {
      type: 'mysql',
      replication: {
        master: { url: main_url },
        slaves: sub_urls.map((url) => ({ url })),
      },
      entities: [join(__dirname, '../**/*.entity.{ts,js}')],
      logging: ['warn', 'error'],
      namingStrategy: new SnakeNamingStrategy(),
      timezone: '+00:00',
    };
  },
});

export const TestConfigs = [
  TestConfig,
  TestMailConfigModule,
  TestTypeOrmConfig,
];

export const getRandomEnumValue = <T>(anEnum: T): T[keyof T] => {
  const enumValues = Object.keys(anEnum) as Array<keyof T>;
  const randomIndex = faker.datatype.number(enumValues.length - 1);
  const randomEnumKey = enumValues[randomIndex];
  return anEnum[randomEnumKey];
};
export const getRandomEnumValues = <T>(anEnum: T): T[keyof T][] => {
  const enumValues = Object.values(anEnum) as Array<keyof T>;
  return faker.helpers.arrayElements(enumValues) as T[keyof T][];
};

export const clearEntities = async (repos: Repository<any>[]) => {
  for (const repo of repos) {
    await repo.query('set foreign_key_checks = 0');
    await repo.clear();
    await repo.query('set foreign_key_checks = 1');
  }
};

export const signInTestUser = async (
  dataSource: DataSource,
  authService: AuthService,
) => {
  const userRepo = dataSource.getRepository(UserEntity);
  const user = await userRepo.save({
    email: faker.internet.email(),
    state: UserStateEnum.Active,
    hashPassword: faker.internet.password(),
  });
  return { jwt: await authService.signIn(UserDto.transform(user)), user };
};

export const DEFAULT_FIELD_COUNT = 2;

export const createQueryBuilder: any = {
  setFindOptions: () =>
    jest.fn().mockImplementation(() => {
      return createQueryBuilder;
    }),
};

export const mockRepository = () => ({
  findOneBy: jest.fn(),
  findOne: jest.fn(),
  find: jest.fn(),
  findBy: jest.fn(),
  findAndCount: jest.fn(),
  findAndCountBy: jest.fn(),
  save: jest.fn(),
  create: jest.fn(),
  update: jest.fn(),
  upsert: jest.fn(),
  count: jest.fn(),
  remove: jest.fn(),
  createQueryBuilder: jest.fn(() => createQueryBuilder),
  query: jest.fn(),
});

export const MockOpensearchRepository = {
  createIndex: jest.fn(),
  deleteIndex: jest.fn(),
  putMappings: jest.fn(),
  createData: jest.fn(),
  getData: jest.fn(),
  updateData: jest.fn(),
  getTotal: jest.fn(),
};
