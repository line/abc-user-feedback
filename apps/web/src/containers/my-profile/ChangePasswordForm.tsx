/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { zodResolver } from '@hookform/resolvers/zod';
import { useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { z } from 'zod';

import { TextInput, toast } from '@ufb/ui';

import { useOAIMutation } from '@/hooks';

type IForm = {
  password: string;
  newPassword: string;
  confirmNewPassword: string;
};
const schema: Zod.ZodType<IForm> = z
  .object({
    password: z.string(),
    newPassword: z.string(),
    confirmNewPassword: z.string(),
  })
  .refine(({ password, newPassword }) => password !== newPassword, {
    message: 'must not equal Password',
    path: ['newPassword'],
  })
  .refine(
    ({ newPassword, confirmNewPassword }) => newPassword === confirmNewPassword,
    {
      message: 'must equal New Password',
      path: ['confirmNewPassword'],
    },
  );

interface IProps extends React.PropsWithChildren {}

const ChangePasswordForm: React.FC<IProps> = () => {
  const { t } = useTranslation();
  const { register, handleSubmit, setError, formState, reset } = useForm<IForm>(
    { resolver: zodResolver(schema) },
  );

  const { mutate, isPending } = useOAIMutation({
    method: 'post',
    path: '/api/users/password/change',
    queryOptions: {
      async onSuccess() {
        toast.positive({ title: t('toast.save') });
        reset({ confirmNewPassword: '', newPassword: '', password: '' });
      },
      onError(error) {
        if (typeof error.message === 'string') {
          if (error.code === 'InvalidPassword') {
            setError('password', { message: 'Invalid Password' });
          }
          toast.negative({ title: error.message as string });
        }
        if (Array.isArray(error.message)) {
          error.message.forEach((message) =>
            toast.negative({ title: message }),
          );
        }
      },
    },
  });

  return (
    <div className="flex flex-col gap-6">
      <div className="flex items-center justify-between">
        <h1 className="font-20-bold">{t('main.profile.change-password')}</h1>
        <button
          form="reset_password"
          className="btn btn-md btn-primary min-w-[120px]"
          disabled={isPending}
        >
          {t('button.save')}
        </button>
      </div>
      <hr />
      <form
        id="reset_password"
        className="flex flex-col gap-6"
        onSubmit={handleSubmit((data) => {
          mutate({ ...data });
        })}
      >
        <TextInput
          type="password"
          label={t('input.label.password')}
          placeholder={t('input.placeholder.password')}
          isSubmitted={formState.isSubmitted}
          isSubmitting={formState.isSubmitting}
          isValid={!formState.errors.password}
          hint={formState.errors.password?.message}
          {...register('password')}
          required
        />
        <TextInput
          type="password"
          label={t('main.profile.label.new-password')}
          placeholder={t('main.profile.placeholder.new-password')}
          isSubmitted={formState.isSubmitted}
          isSubmitting={formState.isSubmitting}
          isValid={!formState.errors.newPassword}
          hint={formState.errors.newPassword?.message}
          {...register('newPassword')}
          required
        />
        <TextInput
          type="password"
          label={t('main.profile.label.confirm-new-password')}
          placeholder={t('main.profile.placeholder.confirm-new-password')}
          isSubmitted={formState.isSubmitted}
          isSubmitting={formState.isSubmitting}
          isValid={!formState.errors.confirmNewPassword}
          hint={formState.errors.confirmNewPassword?.message}
          {...register('confirmNewPassword')}
          required
        />
      </form>
    </div>
  );
};

export default ChangePasswordForm;
