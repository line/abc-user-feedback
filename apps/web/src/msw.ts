/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { http, HttpResponse } from 'msw';
import { setupServer } from 'msw/node';

import { signInWithOAuthMockHandlers } from './features/auth/sign-in-with-oauth/__mocks__/sign-in-with-oauth.mock-handler';
import type {
  OAIMethodPathKeys,
  OAIMethods,
  OAIPathParameters,
} from './shared';
import { convertToColonPath, getRequestUrl } from './shared';

export const server = setupServer(...signInWithOAuthMockHandlers);

export const simpleMockHttp = <
  TMethod extends OAIMethods,
  TPath extends OAIMethodPathKeys<TMethod>,
>({
  method,
  path,
  status = 200,
  params,
  data = {},
}: {
  method: TMethod;
  path: TPath;
  status?: 200 | 201 | 204 | 400 | 401 | 403 | 404 | 500;
  params?: OAIPathParameters<TPath, TMethod>;
  data?: Record<string, unknown>;
}) =>
  server.use(
    http[method](`${convertToColonPath(getRequestUrl(path, params))}`, () =>
      HttpResponse.json(data, { status }),
    ),
  );
