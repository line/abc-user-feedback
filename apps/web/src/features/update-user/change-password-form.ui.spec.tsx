/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { faker } from '@faker-js/faker';
import userEvent from '@testing-library/user-event';

import { simpleMockHttp } from '@/msw';
import { render, screen, waitFor } from '@/test-utils';
import ChangePasswordForm from './change-password-form.ui';

describe('ResetPasswordWithEmailForm', () => {
  // test('match snapshot', () => {
  //   const component = render(<ChangePasswordForm />);
  //   expect(component.container).toMatchSnapshot();
  // });

  // test('validation', async () => {
  //   render(<ChangePasswordForm />);

  //   const saveBtn = screen.getByRole('button', {
  //     name: 'button.save',
  //   });
  //   const passwordInput = screen.getByPlaceholderText(
  //     'input.placeholder.password',
  //   );
  //   const newPasswordInput = screen.getByPlaceholderText(
  //     'main.profile.placeholder.new-password',
  //   );
  //   const confirmPasswordInput = screen.getByPlaceholderText(
  //     'main.profile.placeholder.confirm-new-password',
  //   );

  //   await userEvent.type(passwordInput, faker.string.alphanumeric(8));
  //   await userEvent.type(newPasswordInput, faker.string.alphanumeric(9));
  //   await userEvent.type(confirmPasswordInput, faker.string.alphanumeric(9));

  //   expect(saveBtn).toBeDisabled();

  //   await userEvent.clear(newPasswordInput);
  //   await userEvent.clear(confirmPasswordInput);

  //   const password = faker.string.alphanumeric(8);
  //   await userEvent.type(newPasswordInput, password);
  //   await userEvent.type(confirmPasswordInput, password);

  //   await waitFor(() => expect(saveBtn).not.toBeDisabled());
  // });
  describe('Submittion', () => {
    beforeEach(async () => {
      render(<ChangePasswordForm />);

      const passwordInput = screen.getByPlaceholderText(
        'input.placeholder.password',
      );
      const newPasswordInput = screen.getByPlaceholderText(
        'main.profile.placeholder.new-password',
      );
      const confirmPasswordInput = screen.getByPlaceholderText(
        'main.profile.placeholder.confirm-new-password',
      );
      await userEvent.type(passwordInput, faker.string.alphanumeric(8));

      const password = faker.string.alphanumeric(8);
      await userEvent.type(newPasswordInput, password);
      await userEvent.type(confirmPasswordInput, password);
    });
    test('on Success', async () => {
      simpleMockHttp({
        method: 'post',
        path: '/api/admin/users/password/change',
      });

      const submitBtn = screen.getByRole('button', {
        name: 'button.save',
      });
      await userEvent.click(submitBtn);

      await waitFor(() =>
        expect(
          screen.getByText(new RegExp('toast.save', 'i')),
        ).toBeInTheDocument(),
      );
    });
    // test('on Error', async () => {
    //   simpleMockHttp({
    //     method: 'post',
    //     path: '/api/admin/users/password/change',
    //     status: 500,
    //   });

    //   const submitBtn = screen.getByRole('button', {
    //     name: 'button.save',
    //   });
    //   await userEvent.click(submitBtn);

    //   await waitFor(() =>
    //     expect(screen.getByText(new RegExp('error', 'i'))).toBeInTheDocument(),
    //   );
    // });
  });
});
