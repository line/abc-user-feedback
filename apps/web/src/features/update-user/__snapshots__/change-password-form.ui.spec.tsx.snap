// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPasswordWithEmailForm match snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-6"
  >
    <div
      class="flex items-center justify-between"
    >
      <h1
        class="font-20-bold"
      >
        main.profile.change-password
      </h1>
      <button
        class="btn btn-md btn-primary min-w-[120px]"
        disabled=""
        form="reset_password"
      >
        button.save
      </button>
    </div>
    <hr />
    <form
      class="flex flex-col gap-6"
      id="reset_password"
    >
      <div
        class="flex flex-col gap-2"
      >
        <label
          class="input-label"
          for="input.label.password"
        >
          input.label.password
           
          <span
            class="text-red-primary"
          >
            *
          </span>
        </label>
        <div
          class="relative"
        >
          <input
            class="input  input-md  pr-10 "
            id="input.label.password"
            name="password"
            placeholder="input.placeholder.password"
            required=""
            type="password"
          />
          <div
            class="absolute-y-center absolute right-[14px] flex items-center gap-2"
          />
        </div>
      </div>
      <div
        class="flex flex-col gap-2"
      >
        <label
          class="input-label"
          for="main.profile.label.new-password"
        >
          main.profile.label.new-password
           
          <span
            class="text-red-primary"
          >
            *
          </span>
        </label>
        <div
          class="relative"
        >
          <input
            class="input  input-md  pr-10 "
            id="main.profile.label.new-password"
            name="newPassword"
            placeholder="main.profile.placeholder.new-password"
            required=""
            type="password"
          />
          <div
            class="absolute-y-center absolute right-[14px] flex items-center gap-2"
          />
        </div>
      </div>
      <div
        class="flex flex-col gap-2"
      >
        <label
          class="input-label"
          for="main.profile.label.confirm-new-password"
        >
          main.profile.label.confirm-new-password
           
          <span
            class="text-red-primary"
          >
            *
          </span>
        </label>
        <div
          class="relative"
        >
          <input
            class="input  input-md  pr-10 "
            id="main.profile.label.confirm-new-password"
            name="confirmNewPassword"
            placeholder="main.profile.placeholder.confirm-new-password"
            required=""
            type="password"
          />
          <div
            class="absolute-y-center absolute right-[14px] flex items-center gap-2"
          />
        </div>
      </div>
    </form>
  </div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;
